File: js/array/top-k-frequent-elements.js
/**
 * https://leetcode.com/problems/top-k-frequent-elements
 * 
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
const topKFrequent = (nums, k) => {
    const numsCount = new Map();

    for (num of nums) {
        if (numsCount.has(num)) {
            numsCount.set(num, numsCount.get(num) + 1);
        } else {
            numsCount.set(num, 1);
        }
    };
    return Array.from(numsCount).sort((a, b) => b[1] - a[1]).slice(0, k).map(el => el[0]);
};

//Maybe I can do this???
//Had to learn about prio queues (abstract) and heap implementations
//However my approach is O(n log n), will iterate on this for optimal solution

File: cplusplus/array/top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        priority_queue<pair<int, int>>pq{};
        map<int, int>numFreq{};
        vector<int> res;

        for (auto num : nums) {
            if (numFreq.contains(num)) {
                numFreq.insert({ num, numFreq[num]++ });
            }
            else {
                numFreq.insert({ num, 1 });
            }
        }

        for (auto [key, val] : numFreq) {
            pq.push({ val, key });
        }

        for (int i = 0; i < k; i++) {
            res.push_back(pq.top().second);
            pq.pop();
        }

        return res;
    }
};

// https://leetcode.com/problems/top-k-frequent-elements
// Trick is to use priority queue